src/Chainweb/Miner/Miners.hs

```diff
@@ -96,24 +99,39 @@ submit :<|> poll = client (Proxy :: Proxy MiningAPI)
 -- on a different machine, may be on multiple machines, may be arbitrarily
 -- multithreaded.
 --
-remoteMining :: Manager -> [BaseUrl] -> BlockHeader -> IO BlockHeader
-remoteMining m urls bh = traverseConcurrently_ Par' submission urls >> polling
+-- ASSUMPTION: The contents of the given @NonEmpty BaseUrl@ are unique.
+--
+remoteMining :: Manager -> NonEmpty BaseUrl -> BlockHeader -> IO BlockHeader
+remoteMining m urls bh = submission >> polling
   where
-    -- TODO Better error handling. Don't bail the entire thread if at least one
-    -- miner was successfully communicated with. Just log the rest.
-    submission :: BaseUrl -> IO ()
-    submission url = runClientM (submit bh) (ClientEnv m url Nothing) >>= either throwM pure
+    -- TODO Report /all/ miner calls that errored?
+    -- | Submit work to each given mining client. Will succeed so long as at
+    -- least one call returns back successful.
+    submission :: IO ()
+    submission = do
+        rs <- partitionEithersNEL <$> traverseConcurrently Par' f urls
+        these (throwM . NEL.head) (\_ -> pure ()) (\_ _ -> pure ()) rs
+      where
+        f :: BaseUrl -> IO (Either ServantError ())
+        f url = runClientM (submit bh) $ ClientEnv m url Nothing
 
     -- TODO Use different `Comp`?
     polling :: IO BlockHeader
     polling = fmap head . withScheduler Par' $ \sch ->
         traverse_ (scheduleWork sch . go sch) urls
       where
+        -- TODO Don't bother scheduling retries for `url`s that fail?
         go :: Scheduler IO BlockHeader -> BaseUrl -> IO BlockHeader
-        go sch url = runClientM (poll cid bht) (ClientEnv m url Nothing) >>= \case
-            Left err -> throwM err
-            Right Nothing -> threadDelay 500000 >> go sch url
-            Right (Just new) -> terminateWith sch new
+        go sch url = do
+            -- This prevents scheduled retries from slamming the miners.
+            threadDelay 100000
```

Comment:
I think, this provides another argument to eventually move the logic in the module into an independent process. 

* for competitive production mining (with low level hardware) this interface will probably be to slow / reduce mining margins too much. For Testnet02 DA allows up to 6 blocks per chain per second. On 20 chains that 500ms per block. Polling every 100ms reduce the chances to win a block by up to 20%. This doesn't include the overhead of the reset call itself.

* for scenarios where competitive mining isn't a primary concern (if those are relevant at all), making an additional REST call every 100ms still adds additional networking overhead to consensus, possibly reducing the quality of consensus and causing forks.

A mining coordinator that would use the consensus REST API for communicating with consensus and uses a some custom protocol to communicate with the low-level hardware provides miner with more flexibility to find the best trade-offs for them.

Reply:
Confirming the numbers...

Reply:
Because of the DA "allowance factor" of 3, the fastest (on average) we can expect to mine a block is in 10s (3x faster). The slowest would be 1.5 minutes (3x slower). Ignoring the cost of the call itself, and given that the 100ms wait occurs once **polling** begins (i.e. *after* the work has been submitted), the latest we could find out about the successful mining would be ~99ms after it occurred. Example:

- Millisecond `n`: we checked for a result and got `Nothing`.
- Millsecond `n + 1`: mining succeeds.
- Millisecond n + 100`: we check again, and get a `Just`.

So for the fastest case (10s), the 100ms wait represents a 1% inefficiency in the discovery time. For the slowest case (90s), it represents a 0.11% ineffiency. 

So while it would be strictly more efficient for the miner itself to report its results, the improvement will not be order-of-magnitude.

Reply:
Further, would the miner reporting the results itself back to `chainweb-node` be necessarily faster? Would the handling of that endpoint, plus feeding it back through whatever machinery is necessary to reassociate the new `BlockHeader` with its `Cut` and payload be _less_ than the 100ms here?

Reply:
100ms of course is a made-up number - we can set that to whatever we wish.

Reply:
The mining coordinator would in any case has to the block to the cutdb. Depending on how it's implemented that would be via a Haskell function call or the reset API. So that adds latency, but since its push (and not poll), it should be in the order of microseconds or at most a single digit ms (for REST).

But that's at the level of the mining coordinator. At the level of core miner, I expect that, on the long run, people will use more low-latency protocols with less overhead. Running a web server on mining hardware that has to reply to an request every few (or tens of) seems a bit wasteful.

Reply:
On the other hand I can see how this Rest API could be useful for some intermediate miner, that bundles a number of hardware miners on a box.